### ISSUE TITLE
Optimzer: Better handling of gradients for min/max ops.

### ISSUE THREAD

**Issue by olegserov** [tensorflow/tensorflow#16028]
Please don't kick me too hard for this. If this ticket should not be here, please direct me to the proper place. It's not a help or support request, just a very naive idea/request/improvement.

Image a model has this op in the graph: `y = tf.maximum(a, b)`. If `y` is contributing error to the loss (higher `y` higher is loss), then you are interested in minimizing both `a` and `b`. Otherwise, your optimizer will play whack-a-mole game forever. Especially, if you have something like this in your model: `b = 1 - a` and `y = tf.maximum(a, b)`

I've researching how does optimizer works and looks like each op has a function that defines how to calculate its gradients. 
In the current implementation for maximum/minimum (python's version) is located here: 
https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/ops/math_grad.py#L901
https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/ops/math_grad.py#L883

In the current implementation, it discards gradients for `a`, if `b` is bigger than `a` and other way around. 

I think, that there is some place for improvement of training speed simply by considering gradients for all variables in min/max operations (including reduce_min/reduce_max and all max/min pooling). There are a lot of models with max pooling. It makes sense that they will train faster if they stop playing whack-a-feature each time it has max pooling layer.

If it make sense, I would like to do a small PoC.

**Comment by tensorflowbutler** [comment]
Thank you for your post. We noticed you have not filled out the following field in the issue template. Could you update them if they are relevant in your case, or leave them as N/A? Thanks.
Have I written custom code
OS Platform and Distribution
TensorFlow installed from
TensorFlow version
Bazel version
CUDA/cuDNN version
GPU model and memory
Exact command to reproduce

**Comment by olegserov** [comment]
**Mandatory questions:**
Have I written custom code: N/A? Yes? I have no idea.
OS Platform and Distribution - Any N/A
TensorFlow installed from - pypi
TensorFlow version - '1.4.1'
Bazel version - N/A
CUDA/cuDNN version - N/A
GPU model and memory - N/A
Exact command to reproduce.

**Slow optimizer:**
```
import numpy as np
import tensorflow as tf

a = tf.Variable(10.0)
b = tf.Variable(9.0)
loss = tf.maximum(a, b)

train_step = tf.train.GradientDescentOptimizer(0.5).minimize(loss)

sess = tf.InteractiveSession()

tf.global_variables_initializer().run()


def print_state(sess, step, a, b, loss):
    ret = sess.run({
        'a': a,
        'b': b,
        'loss': loss
    }, feed_dict={})
    
    print ("Step: %3s:\ta=%04.01f\tb=%04.01f\tloss=%04.01f" % (
        step,
        ret['a'],
        ret['b'],
        ret['loss']
    ))


print_state(sess, -1, a, b, loss)

for step in range(10):
    sess.run(train_step, feed_dict={})

    print_state(sess, step, a, b, loss)

```
Outputs:
```
Step:  -1:	a=10.0	b=09.0	loss=10.0
Step:   0:	a=09.5	b=09.0	loss=09.5
Step:   1:	a=09.0	b=09.0	loss=09.0
Step:   2:	a=08.5	b=09.0	loss=09.0
Step:   3:	a=08.5	b=08.5	loss=08.5
Step:   4:	a=08.0	b=08.5	loss=08.5
Step:   5:	a=08.0	b=08.0	loss=08.0
Step:   6:	a=07.5	b=08.0	loss=08.0
Step:   7:	a=07.5	b=07.5	loss=07.5
Step:   8:	a=07.0	b=07.5	loss=07.5
Step:   9:	a=07.0	b=07.0	loss=07.0
```

As you can see it is pooling `a` down, then `b`, and then again. It will require less steps if it is going to change both `a` and `b` together.

**Optimizer playing whack-a-feature game**
```
import numpy as np
import tensorflow as tf

a = tf.Variable(10.0)
b = 10.0 - a
loss = tf.maximum(a, b)

train_step = tf.train.GradientDescentOptimizer(0.5).minimize(loss)

sess = tf.InteractiveSession()

tf.global_variables_initializer().run()


def print_state(sess, step, a, b, loss):
    ret = sess.run({
        'a': a,
        'b': b,
        'loss': loss
    }, feed_dict={})
    
    print ("Step: %3s:\ta=%04.01f\tb=%04.01f\tloss=%04.01f" % (
        step,
        ret['a'],
        ret['b'],
        ret['loss']
    ))


print_state(sess, -1, a, b, loss)

for step in range(20):
    sess.run(train_step, feed_dict={})

    print_state(sess, step, a, b, loss)
```

```
Step:  -1:	a=10.0	b=00.0	loss=10.0
Step:   0:	a=09.5	b=00.5	loss=09.5
Step:   1:	a=09.0	b=01.0	loss=09.0
Step:   2:	a=08.5	b=01.5	loss=08.5
Step:   3:	a=08.0	b=02.0	loss=08.0
Step:   4:	a=07.5	b=02.5	loss=07.5
Step:   5:	a=07.0	b=03.0	loss=07.0
Step:   6:	a=06.5	b=03.5	loss=06.5
Step:   7:	a=06.0	b=04.0	loss=06.0
Step:   8:	a=05.5	b=04.5	loss=05.5
Step:   9:	a=05.0	b=05.0	loss=05.0
Step:  10:	a=04.5	b=05.5	loss=05.5
Step:  11:	a=05.0	b=05.0	loss=05.0
Step:  12:	a=04.5	b=05.5	loss=05.5
Step:  13:	a=05.0	b=05.0	loss=05.0
Step:  14:	a=04.5	b=05.5	loss=05.5
Step:  15:	a=05.0	b=05.0	loss=05.0
Step:  16:	a=04.5	b=05.5	loss=05.5
Step:  17:	a=05.0	b=05.0	loss=05.0
Step:  18:	a=04.5	b=05.5	loss=05.5
Step:  19:	a=05.0	b=05.0	loss=05.0
```

As you can see, it is stuck in a cycle. If it considered loss for both `a` and `b` it would have quickly come to the equilibrium.

**Comment by tensorflowbutler** [comment]
The original poster has replied to this issue after the stat:awaiting response label was applied.

**Comment by drpngx** [comment]
So the canonical way to tackle this problem is probably to "soften" the non-linear function, e.g use something like:
```
a_is_bigger = tf.sigmoid((a-b) / width)
y = a_is_bigger * a + (1 - a_is_bigger) * b
```
and increase `width` over time.

The second canonical way is to use an optimizer that implements the proximal [gradients](http://www.stat.cmu.edu/~ryantibs/convexopt-S15/lectures/08-prox-grad.pdf).

If you have a practical solution, we're always open to good algorithms! In general if it's been tested on a competitive benchmark and it works, it would be great to contribute.

**Comment by julien2512** [comment]
@drpngx Should we use something like that 

> http://163.172.41.53:8080/#activation=tanh&batchSize=10&dataset=circle&regDataset=reg-maximum&learningRate=0.0001&regularizationRate=0&noise=0&networkShape=3,2&seed=0.29413&showTestData=false&discretize=false&percTrainData=50&x=true&y=true&xTimesY=false&xSquared=false&ySquared=false&cosX=false&sinX=false&cosY=false&sinY=false&collectStats=false&problem=regression&initZero=false&hideText=false

to implement the maximum function in order the gradient to work (with static values rather than variables) ?
Width seems not to be necessary this way.

**Comment by julien2512** [comment]
I am not confortable with the "width" parameter in
`a_is_bigger = tf.sigmoid((a-b) / width)`

So I tried to keep your main idea in a different form (there it is 2 hidden layer network using tanh activation to emulate the maximum function).
I am on it right now.

**Comment by julien2512** [comment]
My method only works with a given radius (a and b between [-radius, radius] for example).
Also, their might be biases near 0 on both a and b.

I got bad results by now. May be I missed something.

**Comment by julien2512** [comment]
Alright, I got it. I was looking for something to much complicated : a single hidden layer is enough.

I have soften argmax that way
V (a,b) C [-6,6]² ,  argmax(a,b) = 0,5× (tanh (-6,7×a+6,7×b−5))+0,5
That's what I am going to test.

Soften max is a bit harder but seem to be possible a similar way (with more hidden layers).

**Comment by julien2512** [comment]
The derivative of argmax at the point 0 will then be 1.
Applying a gradient when a=b will lead to an infinite loop (we can't decide which direction to take).

tf.maximum will have the same problem. We can't find an answer with the tf.argmax or tf.max only I guess.

**Comment by julien2512** [comment]
That is not about the value we want to get, but how we want it to behave.

**Comment by julien2512** [comment]
@drpngx now I understood your proposal :
```
a_is_bigger = tf.sigmoid((a-b) / width)
y = a_is_bigger * a + (1 - a_is_bigger) * b
```
When a is close to b, the growing parameter width will increase the difference between a and b to make a_is_bigger be 1 or 0.

What do you think about
`width = 10^(floor(log(abs(a-b))))`
?

I do not know about floor and abs gradient thus.

**Comment by julien2512** [comment]
https://www.desmos.com/calculator/dwiaagflr3

It still do not work near 0. The derivative of a_is_bigger is just too high.

**Comment by julien2512** [comment]
The code for current max (and min !) gradient : 
```
def _MinOrMaxGrad(op, grad):
  """Gradient for Min or Max. Amazingly it's precisely the same code."""
  input_shape = array_ops.shape(op.inputs[0])
  output_shape_kept_dims = math_ops.reduced_shape(input_shape, op.inputs[1])
  y = op.outputs[0]
  y = array_ops.reshape(y, output_shape_kept_dims)
  grad = array_ops.reshape(grad, output_shape_kept_dims)

  # Compute the number of selected (maximum or minimum) elements in each
  # reduction dimension. If there are multiple minimum or maximum elements
  # then the gradient will be divided between them.
  indicators = math_ops.cast(math_ops.equal(y, op.inputs[0]), grad.dtype)
  num_selected = array_ops.reshape(
      math_ops.reduce_sum(indicators, op.inputs[1]), output_shape_kept_dims)

return [math_ops.div(indicators, num_selected) * grad, None]
```

**Comment by julien2512** [comment]
Gradient for abs :
```
@ops.RegisterGradient("Abs")
def _AbsGrad(op, grad):
  x = op.inputs[0]
return grad * math_ops.sign(x)
```

but floor is not implemented :
```
@ops.RegisterGradient("Floor")
def _FloorGrad(_, unused_grad):
return [None]
```

**Comment by julien2512** [comment]
For all of these reason, despite the tentation to soften max (nor argmax) I think I will use tf.conf instead.
I keep you in touch with my results.

**Comment by ajay-anand** [comment]
Just a comment,

Let y = max( a, b) such that a >> b,
then we can approximate it as 
y = log(e^a + e^b)
If we compute the gradient with respect to x, 
∂y/∂x = ∂a/∂x e^a / (e^a+e^b) + ∂b/∂x e^b / (e^a+e^b)
Notice that with the assumption a >> b, the first term approaches to ∂a/∂x and the second term vanishes. This means if a >> b, it is enough to propagate gradients and update parameters associated only with **a** whenever expression max(a,b) appears. 

If the assumption that (a >> b) or (b >> a) is true, only then max(a, b) can be converted to a continuous function as done above. Otherwise we need another measure, say a distance measure to approximate non-continuous max with a  continuous function in a fuzzy sense. This 'distance' in essence is a fuzzification parameter that will help relatively define whether a and b are 'similar' or 'way too different'. But this distance measure depends on the nature and distribution of data, so it has to be either supplied by the user or discovered from the data itself (if at all a broader view of data is available).

**Comment by julien2512** [comment]
I am agree, I realized it was better to use a custom gradient.

**Comment by iiphii** [comment]
I haven't tried this myself but the concept of Lp-norm seems to me like a natural approach for "softening" the max function. For your two number case, the Lp-norm is defined as 
`y = (|a|^p + |b|^p)^(1/p)`
where p is a free parameter.

The case p->∞ corresponds to
`y = tf.maximum(a,b)`
and the case p=2 corresponds to the familiar Eucledian distance. Perhaps there could be an intermediate p value which would suit your use case.

**Comment by tilakrayal** [comment]
@olegserov,
Could you please take a look at this [comment](https://github.com/tensorflow/tensorflow/issues/16028#issuecomment-423733796) from the user and  kindly update to the latest stable version 2.10 and let us know if you are facing the same issue. Thank you!
